dmoj/judge.py

```diff
@@ -175,41 +179,202 @@ def begin_grading(
+        conn.recv()
+        self._terminate_grading = True
+        if self.grader:
+            self.grader.terminate_grading()
+
+    def _run_grading_wrapper_process(self, conn):
+        try:
+            # TODO(tbrindus): set different process title?
+
+            # TODO(tbrindus): name all these threads for exception readability
+            threading.Thread(target=self._run_abort_request_handler_thread, args=(conn,)).start()
+
+            for type, data in self._block_and_grade_cases():
+                conn.send((type, data))
+        except:  # noqa: E722, we explicitly want to notify the parent of everything
+            exc = sys.exc_info()
+            message = ''.join(traceback.format_exception(*exc))
+            conn.send((IPC_UNHANDLED_EXCEPTION, message))
+
+        # Signal to the other end that we're done.
+        conn.send(None)
+
+    def _block_and_grade_cases(self):
+        try:
+            problem = Problem(self.problem_id, self.time_limit, self.memory_limit, self.meta)
+
+            if 'signature_grader' in problem.config:
+                grader_class = graders.SignatureGrader
+            elif 'interactive' in problem.config:
+                grader_class = graders.BridgedInteractiveGrader
+            elif 'custom_judge' in problem.config:
+                grader_class = graders.CustomGrader
+            else:
+                grader_class = graders.StandardGrader
+
+            try:
+                self.grader = grader_class(self, problem, self.language, utf8bytes(self.source))
+            except CompileError as compilation_error:
+                error = compilation_error.args[0] or b'compiler exited abnormally'
+                yield IPC_COMPILE_ERROR, error
+                return
+            else:
+                binary = self.grader.binary
+                if hasattr(binary, 'warning') and binary.warning is not None:
+                    yield IPC_COMPILE_MESSAGE, binary.warning
+
+            yield IPC_GRADING_BEGIN, self.grader.is_pretested
+
+            batch_counter = 1
+
+            # cases are indexed at 1
+            case_number = 1
+            try:
+                for result in self._grade_cases(self.grader, self.grader.cases(), short_circuit=self.short_circuit):
+                    if isinstance(result, SubmissionJudgeProcess.BatchBegin):
+                        yield IPC_BATCH_BEGIN, batch_counter
+                    elif isinstance(result, SubmissionJudgeProcess.BatchEnd):
+                        yield IPC_BATCH_END, batch_counter
+                        batch_counter += 1
+                    else:
+                        # TODO(tbrindus): we can't pickle TestCase
+                        yield IPC_RESULT, (case_number, result)
+                        case_number += 1
+            except TerminateGrading:
+                yield IPC_GRADING_TERMINATED, None
+            else:
+                yield IPC_GRADING_END, None
+        except GeneratorExit:
```

Comment:
Why do we need to catch `GeneratorExit`?

Reply:
Fixed.

Reply:
In case the consumer of the generator fails during `conn.send`, I had this crash on me.

Reply:
I've removed this since other changes made it unnecessary, I think.